rfcs/convergence/make-overrides.md

```diff
@@ -0,0 +1,195 @@
+  const iconClasses = useIconStyles();
+
+  return (
+    <div className={ax(rootClasses, props.className)}>
+      {props.icon && <div className={ax(iconClasses, props.icon.className)} />}
+    </div>
+  );
+}
+```
+
+To compute classnames we need to understand what styles should be applied and thus execute matchers on each render as memoization of user's input is more expensive.
+
+This moves us to a next issue: there is no way to define styles for multiple slots/components with a single call of the `makeStyles()`. Each slot will require a separate call of `makeStyles()`. Since these calls are represented by React hooks we can't bail out even if these slots  are not rendered.
+
+_Side note:_
+
+> Initially we have been focused on implementation of atomic CSS and merging style definitions. As a source for inspirations we have used [Facebook stylex talk](https://www.youtube.com/watch?v=9JZHodNR184):
+>
+> - there was nothing like [`ax()`](https://github.com/microsoft/fluentui/pull/16411) function to merge atomic classes
+> - there was no contextual RTL support (parts of app can be rendered with different text directions)
+
+## Detailed Design or Proposal
+
+To solve these issues we made a step back to the original API of `makeStyles()` (current iteration is a second version) and as we introduced `ax()` to merge atomic classnames we explored different solutions.
+
+Proposed API solve the problem in a "vandal" way 🪓 Matchers are moved to user's scope thus we can have all definitions in a single `makeStyles()` call => we have a single React hook. See a modified snippet below:
+
+```tsx
+import { ax, makeOverrides } from "@fluentui/react-make-styles";
+
+const useStyles = makeStyles({
+  /* 👍 no matchers, no need to execute on each render */
+  root: { color: "red" },
+  rootPrimary: { color: "blue" },
+
+  /* 👍 styles for each slot are defined together (not a requirement) */
+  icon: { background: "black" },
+  iconPrimary: { background: "white" }
+});
+
+function Component() {
+  /* 👍 a single call of React hook */
+  const classes = useStyles();
+
+  return (
+    <div
+      className={ax(
+        classes.root /* The concept of matching is replaced with selective classname concat */,
+        props.primary && classes.rootPrimary,
+        props.className
+      )}
+    >
+      {props.icon && (
+        <div
+          className={ax(
+            classes.icon,
+            props.primary && classes.iconPrimary,
+            props.icon.className
+          )}
+        />
+      )}
+    </div>
+  );
+}
+```
+
+### Pros and Cons
+
+These changes result in performance improvements in `makeOveriddes`:
+
+- matcher functions are no longer executed on each render in styling functions
+- styles are executed and merged only when slots are present
+- less React hooks to call for slots
+
+Surprisingly, this also moves us closer to [CSS Modules](https://github.com/css-modules/css-modules). The same snippet with CSS Modules, for example:
```

Comment:
So, does this change make it more feasible to have versions of our components that use CSS Modules or static styling?

Reply:
From API point of view: yes, it looks pretty close to CSS Modules (actually, that's why I made this snippet)
From implementation point of view: no, as `makeStyles()`/`makeOverrides()` are using atomic CSS